src/structures/ModalSubmitInteraction.js

```diff
@@ -0,0 +1,111 @@
+'use strict';
+
+const Interaction = require('./Interaction');
+const InteractionWebhook = require('./InteractionWebhook');
+const ModalSubmitFieldsResolver = require('./ModalSubmitFieldsResolver');
+const InteractionResponses = require('./interfaces/InteractionResponses');
+const { MessageComponentTypes } = require('../util/Constants');
+
+/**
+ * Represents a modal submit interaction.
+ * @extends {Interaction}
+ * @implements {InteractionResponses}
+ */
+class ModalSubmitInteraction extends Interaction {
+  constructor(client, data) {
+    super(client, data);
+
+    /**
+     * The custom id of the modal.
+     * @type {string}
+     */
+    this.customId = data.data.custom_id;
+
+    /**
+     * @typedef {Object} PartialTextInputData
+     * @property {string} [customId] A unique string to be sent in the interaction when submitted
+     * @property {MessageComponentType} [type] The type of this component
+     * @property {string} [value] Value of this text input component
+     */
+
+    /**
+     * @typedef {Object} PartialModalActionRow
+     * @property {MessageComponentType} [type] The type of this component
+     * @property {PartialTextInputData[]} [components] Partial text input components
+     */
+
+    /**
+     * The inputs within the modal
+     * @type {PartialModalActionRow[]}
+     */
+    this.components =
+      data.data.components?.map(c => ({
+        type: MessageComponentTypes[c.type],
+        components: ModalSubmitInteraction.transformComponent(c),
+      })) ?? [];
+
+    /**
+     * The message associated with this interaction
+     * @type {Message|APIMessage|null}
+     */
+    this.message = this.channel?.messages._add(data.message) ?? data.message ?? null;
```

Comment:
i just ran into the same issue. trace:

```sh
TypeError: Cannot read properties of undefined (reading 'id')
    at MessageManager._add (/service/node_modules/discord.js/src/managers/CachedManager.js:47:48)
    at MessageManager._add (/service/node_modules/discord.js/src/managers/MessageManager.js:32:18)
    at new ModalSubmitInteraction (/service/node_modules/discord.js/src/structures/ModalSubmitInteraction.js:51:43)
    at InteractionCreateAction.handle (/service/node_modules/discord.js/src/client/actions/InteractionCreate.js:71:25)
    at Object.module.exports [as INTERACTION_CREATE] (/service/node_modules/discord.js/src/client/websocket/handlers/INTERACTION_CREATE.js:4:36)
    at WebSocketManager.handlePacket (/service/node_modules/discord.js/src/client/websocket/WebSocketManager.js:351:31)
    at WebSocketShard.onPacket (/service/node_modules/discord.js/src/client/websocket/WebSocketShard.js:444:22)
    at WebSocketShard.onMessage (/service/node_modules/discord.js/src/client/websocket/WebSocketShard.js:301:10)
    at WebSocket.onMessage (/service/node_modules/discord.js/node_modules/ws/lib/event-target.js:199:18)
    at WebSocket.emit (node:events:390:28)
```

what is the suggested usage pattern @monbrey? i was trying to show a modal in response to a command interaction to gather user input. basic reproduction pseudocode:

1. /test
2. showModal
3. awaitModalSubmit <-- crash

Reply:
Just wait until I fix it

Reply:
```js
this.message = data.message ? this.channel?.messages._add(data.message) : (data.message ?? null);
```

this fixes it (and i believe preserves the same logic). however, ultimately it seems to need a message.

> `DiscordAPIError: Interaction has already been acknowledged.`

so in some ways the `showModal` _counts_ as an interaction response (acknowledgement?). however, without some ref to a message it breaks the downstream replies.

so i think in order to fix this further (at least for initiating in a command interaction) it will need to somehow mark the interaction as `replied` (or `deferred` may be more appropriate?).

Reply:
for the record im not a fan of the one-liner "fix" i put here but i was just trying to stay with the code style.

Reply:
i tried out the theory by adding to `showModal` implementation in https://github.com/monbrey/discord.js/blob/fcb26ad46b9ce5fc0033dba211ec7c0bd2a5f6d2/src/structures/interfaces/InteractionResponses.js#L234-L242

```js
  async showModal(modal) {
    const _modal = modal instanceof Modal ? modal : new Modal(modal);
// ++
    this.replied = true;
//
    await this.client.api.interactions(this.id, this.token).callback.post({
      data: {
        type: InteractionResponseTypes.MODAL,
        data: _modal.toJSON(),
      },
    });
  }
```

this **does fix** the issue but now the modal does not close when submitted. no errors are reported in DJS so i dont know how to debug further.

Reply:
There's a lot to unpack here, some correct, some incorrect.

The one liner isn't quite correct. `data.message` can't be on the right of the `:`, that's where its a falsey value.
```js
this.message = data.message ? this.channel?.messages._add(data.message) ?? data.message : null;
```

Correct - showing a modal is counted as a response to an interaction, and I should be setting `this.replied` when it's done. I might need to check on the v14 PR for this also, I'm not sure it was being set there either.